drivers/pwm/pwm_led_esp32.c

```diff
@@ -0,0 +1,517 @@
+
+	sys_set_bit(conf0_addr, LEDC_SIG_OUT_EN_LSCH0_S);
+	sys_set_bit(conf1_addr, LEDC_DUTY_START_LSCH0_S);
+
+	pwm_led_esp32_low_speed_update(speed_mode, channel);
+}
+
+static void pwm_led_esp32_duty_config(int speed_mode,
+				      int channel,
+				      int duty_val,
+				      union pwm_led_esp32_duty duty)
+{
+	volatile u32_t hpoint_addr;
+	volatile u32_t duty_addr;
+	volatile u32_t conf1_addr;
+	volatile u32_t conf1_val;
+
+	if (speed_mode == PWM_LED_ESP32_HIGH_SPEED) {
+		hpoint_addr = PWM_ESP32_HSCH_HPOINT(channel);
+		duty_addr = PWM_ESP32_HSCH_DUTY(channel);
+		conf1_addr = PWM_ESP32_HSCH_CONF1(channel);
+
+	} else {
+		hpoint_addr = PWM_ESP32_LSCH_HPOINT(channel);
+		duty_addr = PWM_ESP32_LSCH_DUTY(channel);
+		conf1_addr = PWM_ESP32_LSCH_CONF1(channel);
+	}
+
+	sys_write32(0, hpoint_addr);
+	sys_write32(duty_val, duty_addr);
+	sys_write32(duty.val, conf1_addr);
+
+	pwm_led_esp32_low_speed_update(speed_mode, channel);
+}
+
+static void pwm_led_esp32_duty_set(int speed_mode, int channel, int duty_val)
+{
+	union pwm_led_esp32_duty duty;
+
+	duty.start = 0;
+	duty.direction = 1;
+	duty.cycle = 1;
+	duty.scale = 0;
+
+	pwm_led_esp32_duty_config(speed_mode, channel, duty_val << 4, duty);
+}
+
+static void pwm_led_esp32_bind_channel_timer(int speed_mode,
+					     int channel,
+					     int timer)
+{
+	volatile u32_t timer_addr;
+
+	if (speed_mode == PWM_LED_ESP32_HIGH_SPEED) {
+		timer_addr = PWM_ESP32_HSCH_CONF0(channel);
+	} else {
+		timer_addr = PWM_ESP32_LSCH_CONF0(channel);
+	}
+
+	set_mask32(timer, timer_addr);
+
+	pwm_led_esp32_low_speed_update(speed_mode, channel);
+}
+
+static int pwm_led_esp32_channel_set(int pin, bool speed_mode, int channel,
+				     int duty, int timer)
+{
+	const int pin_mode = GPIO_DIR_OUT;
+
+	const char *device_name;
+	struct device *gpio;
+	int ret;
+	u32_t sig_out_idx;
+
+	/* Set duty cycle */
+	pwm_led_esp32_duty_set(speed_mode, channel, duty);
+
+	pwm_led_esp32_update_duty(speed_mode, channel);
+
+	pwm_led_esp32_bind_channel_timer(speed_mode, channel, timer);
+
+	/* Set pin */
+	device_name = esp32_get_gpio_for_pin(pin);
+	if (!device_name) {
+		return -EINVAL;
+	}
+
+	gpio = device_get_binding(device_name);
+	if (!gpio) {
+		return -EINVAL;
+	}
+
+	ret = gpio_pin_configure(gpio, pin, pin_mode);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (speed_mode == PWM_LED_ESP32_HIGH_SPEED) {
+		sig_out_idx = LEDC_HS_SIG_OUT0_IDX + channel;
+	} else {
+		sig_out_idx = LEDC_LS_SIG_OUT0_IDX + channel;
+	}
+	esp32_rom_gpio_matrix_out(pin, sig_out_idx, 0, 0);
+
+	return 0;
+}
+
+static int pwm_led_esp32_timer_set(int speed_mode, int timer,
+				   int bit_num, int frequency)
+{
+	u32_t timer_addr;
+	u64_t div_num;
+	int tick_sel = PWM_LED_ESP32_APB_CLK_FREQ;
+	u32_t precision = (0x1 << bit_num);
+
+	/* This expression comes from ESP32 Espressif's Technical Reference
+	 * Manual chapter 13.2.2 Timers.
+	 * div_num is a fixed point value (Q10.8).
+	 */
+	div_num = ((u64_t) APB_CLK_FREQ << 8) / frequency / precision;
```

Comment:
Please add an assertion for frequency and precision here, to have better diagnostic if someone passes 0 to these parameters.

Reply:
I'll add assertion only for frequency. Precision is never 0 because this: 
`u32_t precision = (0x1 << bit_num);`